smithy-typescript-codegen/src/main/java/software/amazon/smithy/typescript/codegen/integration/HttpBindingProtocolGenerator.java

```diff
@@ -519,6 +521,20 @@ private void generateOperationResponseSerializer(
         }
     }
 
+    private void calculateContentLength(GenerationContext context) {
+        TypeScriptWriter writer = context.getWriter();
+        writer.addDependency(TypeScriptDependency.AWS_SDK_UTIL_BODY_LENGTH_NODE);
+        writer.addImport("calculateBodyLength", null, "@aws-sdk/util-body-length-node");
+        writer.openBlock("if (body && Object.keys(headers).map((str) => str.toLowerCase())"
+                + ".indexOf('content-length') === -1) {", "}", () -> {
+            writer.write("const length = calculateBodyLength(body);");
+            writer.openBlock("if (length !== undefined) {", "}", () -> {
+                writer.write("headers = { ...headers, 'content-length': String(length) };");
```

Comment:
why do we reassign headers instead of mutating it?

Reply:
I based this off of [what the client middleware plugin does](https://github.com/aws/aws-sdk-js-v3/blob/main/packages/middleware-content-length/src/index.ts#L29-L32). I'm not terribly attached to this approach though, so I can change it.

Reply:
The reassigning was done to usable immutability. We'd discovered a bug in during pre-GA because of mutation, which was fixed by using destructuring assignment